### PR TITLE
Refactor `Strategy._move_optimizer_states` as utility functions

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Refactoring to make the code easier to test by splitting this out as a separate function. I didn't see any existing tests for this. This will simplify https://github.com/PyTorchLightning/pytorch-lightning/pull/11757

### Does your PR introduce any breaking changes? If yes, please list them.

Yes, if custom Strategy implementations overrode `_move_optimizer_to_device`, they will break after this PR

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@ananthsub do you think we could unify this with some similar lines of code we have in `CheckpointConnector.restore_optimizers`?

**Comment by ananthsub** [comment]
> @ananthsub do you think we could unify this with some similar lines of code we have in `CheckpointConnector.restore_optimizers`?

Yes exactly. that was the motivation after seeing the bug here: https://github.com/PyTorchLightning/pytorch-lightning/issues/11741

See my comment here: https://github.com/PyTorchLightning/pytorch-lightning/pull/11757#discussion_r800075578

File: pytorch_lightning/utilities/optimizer.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Copyright The PyTorch Lightning team.
```
Comment (awaelchli) [inline pytorch_lightning/utilities/optimizer.py:1]: would it perhaps better fit into the already existing apply_func.py, where we have other utilities to transfer collections of data from device to device?

File: tests/utilities/test_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import collections
+
+import torch
+
+from pytorch_lightning.utilities.optimizer import optimizer_to_device
+
+
+def test_optimizer_to_device():
+    """Ensure that after the initial seed everything, the seed stays the same for the same run."""
```
Comment (awaelchli) [inline tests/utilities/test_optimizer.py:9]: the docstring appears to be from a copy-paste of an unrelated test

File: pytorch_lightning/utilities/optimizer.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/utilities/optimizer.py:1]: the reason i added a new file with optimizers in the name was that it's not obvious `apply_func` would contain this sort of utility, which makes it harder for new users to find this.

i could move it to `apply_func` if you feel strongly about it.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -279,17 +279,6 @@ def restore_optimizers(self) -> None:
...
-            # move optimizer to GPU 1 weight at a time
-            # avoids OOM
-            if self.trainer.root_gpu is not None:
-                for param, state in optimizer.state.items():
-                    if isinstance(state, dict):
-                        for k, v in state.items():
-                            if isinstance(v, torch.Tensor):
-                                state[k] = v.cuda(self.trainer.root_gpu)
-                    elif isinstance(state, torch.Tensor):
-                        optimizer.state[param] = state.cuda(self.trainer.root_gpu)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:292]: @SeanNaren I see the DeepSpeed strategy overrides `load_optimizer_state_dict` to do nothing

After this change, do the optimizer states now need to be moved to the root_device within the strategy? Or does  the DeepSpeed engine internally already handle this?

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:292]: @ananthsub The test `test_deepspeed_multigpu_stage_3_warns_resume_training` should fail if the optimizer states where not moved to the root_device ? Can you make sure of this ?

Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:292]: I had a brief look at the deepspeed code and didn't see anything that would break with these changes, however as @tchaton suggested I think the tests will verify this.

File: pytorch_lightning/strategies/strategy.py [hunk 4]
```diff
@@ -140,14 +141,6 @@ def setup_precision_plugin(self) -> None:
         self.optimizers = optimizers
         self.lr_scheduler_configs = lr_scheduler_configs
 
-    def _move_optimizer_state(self, device: Optional[torch.device] = None) -> None:
-        """Moves the state of the optimizers to the appropriate device if needed."""
-        for opt in self.optimizers:
-            for p, v in opt.state.items():
-                # `self.root_device` would raise error if called outside the spawn process
-                # while training on 8 and more cores.
```
Comment (awaelchli) [inline pytorch_lightning/strategies/strategy.py:148]: this is solved I guess, since we pulled up the process spawning to the beginning of Trainer.fit()

**Review by awaelchli — APPROVED** [review]
